proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+        PropagationData,
+
+        /// The activity object should be populated with all the propagation info and also all other
+        /// properties such as Links, Tags, and Events. Activity.IsAllDataRequested will return true.
+        AllData,
+
+        /// The activity object should be populated the same as the AllData case and additionally
+        /// Activity.IsRecorded is set true. For activities using W3C trace ids this sets a flag bit in the
+        /// ID that will be propagated downstream requesting that trace is recorded everywhere.
+        AllDataAndRecorded
+    }
+}
+```
+
+### ActivityCreationOptions
+
+```c#
+namespace System.Diagnostics
+{
+    public readonly struct ActivityCreationOptions<T>
+    {
+        public ActivitySource Source { get; }
+        public string Name { get; }
+        public ActivityKind Kind { get; }
+        public T Parent { get; }
+        public IEnumerable<KeyValuePair<string, string?>> Tags { get; }
+        public IEnumerable<ActivityLink> Links { get; }
+    }
+}
+```
+
+### ActivityListener
+
+```C#
+namespace System.Diagnostics
+{
+    public delegate ActivityDataRequest GetRequestedData<T>(ref ActivityCreationOptions<T> options);
+
+    public sealed class ActivityListener : IDisposable
+    {
+        public ActivityListener() { }
+        public Action<Activity> ActivityStopped { get; set; }
+        public Action<Activity> ActivityStarted { get; set; }
+        public Func<ActivitySource, bool> ShouldListenTo { get; set; }
+        public GetRequestedData<string> GetRequestedDataUsingParentId { get; set; }
+        public GetRequestedData<ActivityContext> GetRequestedDataUsingContext { get; set; }
+        public void Dispose() { }
+    }    
+}
+```
+
+### ActivitySource
+
+```C#
+namespace System.Diagnostics
+{
+    /// ActivitySource is the type that help create and start Activity and allow listening to the Activity events too.
+    public sealed class ActivitySource : IDisposable
+    {
+        /// Construct an ActivitySource object with the input name and component version
+        public ActivitySource(string name, string version = "");
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public string Version { get; }
+
+        /// Check if there is any listeners for this ActivitySource.
+        /// This property can be helpful to tell if there is no listener, then no need to create Activity object
+        /// and avoid creating the objects needed to create Activity (e.g. ActivityContext)
+        /// Example of that is http scenario which can avoid reading the context data from the wire.
+        public bool HasListeners();
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
```

Comment:
Does this method starts the activity and also sets it as the active one? i.e will Activity.Current be set to this newly created Activity?

OT spec states Span creation should NOT set the span as the active one.

_"Span creation MUST NOT set the newly created Span as the currently active Span by default, but this functionality MAY be offered additionally as a separate operation"_

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span-creation

Reply:
> Does this method starts the activity and also sets it as the active one? i.e will Activity.Current be set to this newly created Activity?

Yes.

> OT spec states Span creation should NOT set the span as the active one.

You still have the option to do `new Activity(...)` as before if you really want not to set the activity as current. Mainstream scenarios always want to set the Activity as current when creating it.

Reply:
okay.
Its good to be aware of the 2 notable differences from the OT Spec
1. StartActivity - starts the Activity and makes it active. (OT says starting and activating are separate)
2. new Activity() can be used to create but not activate Activity. (OT says Spans should be allowed to created only from Tracer.)

Not blocker, just calling out so as to help others reading this after me :)